node-js-basics-101/node_modules/socket.io-adapter/index.js

```diff
@@ -0,0 +1,263 @@
+
+/**
+ * Adds a socket to a room.
+ *
+ * @param {String} socket id
+ * @param {String} room name
+ * @param {Function} callback
+ * @api public
+ */
+
+Adapter.prototype.add = function(id, room, fn){
+  return this.addAll(id, [ room ], fn);
+};
+
+/**
+ * Adds a socket to a list of room.
+ *
+ * @param {String} socket id
+ * @param {String} rooms
+ * @param {Function} callback
+ * @api public
+ */
+
+Adapter.prototype.addAll = function(id, rooms, fn){
+  for (var i = 0; i < rooms.length; i++) {
+    var room = rooms[i];
+    this.sids[id] = this.sids[id] || {};
+    this.sids[id][room] = true;
+    this.rooms[room] = this.rooms[room] || Room();
+    this.rooms[room].add(id);
+  }
+  if (fn) process.nextTick(fn.bind(null, null));
+};
+
+/**
+ * Removes a socket from a room.
+ *
+ * @param {String} socket id
+ * @param {String} room name
+ * @param {Function} callback
+ * @api public
+ */
+
+Adapter.prototype.del = function(id, room, fn){
+  this.sids[id] = this.sids[id] || {};
+  delete this.sids[id][room];
+  if (this.rooms.hasOwnProperty(room)) {
+    this.rooms[room].del(id);
+    if (this.rooms[room].length === 0) delete this.rooms[room];
+  }
+
+  if (fn) process.nextTick(fn.bind(null, null));
+};
+
+/**
+ * Removes a socket from all rooms it's joined.
+ *
+ * @param {String} socket id
+ * @param {Function} callback
+ * @api public
+ */
+
+Adapter.prototype.delAll = function(id, fn){
+  var rooms = this.sids[id];
+  if (rooms) {
+    for (var room in rooms) {
+      if (this.rooms.hasOwnProperty(room)) {
+        this.rooms[room].del(id);
+        if (this.rooms[room].length === 0) delete this.rooms[room];
+      }
+    }
+  }
+  delete this.sids[id];
+
+  if (fn) process.nextTick(fn.bind(null, null));
+};
+
+/**
+ * Broadcasts a packet.
+ *
+ * Options:
+ *  - `flags` {Object} flags for this packet
+ *  - `except` {Array} sids that should be excluded
+ *  - `rooms` {Array} list of rooms to broadcast to
+ *
+ * @param {Object} packet object
+ * @api public
+ */
+
+Adapter.prototype.broadcast = function(packet, opts){
+  var rooms = opts.rooms || [];
+  var except = opts.except || [];
+  var flags = opts.flags || {};
+  var packetOpts = {
+    preEncoded: true,
+    volatile: flags.volatile,
+    compress: flags.compress
+  };
+  var ids = {};
+  var self = this;
+  var socket;
+
+  packet.nsp = this.nsp.name;
+  this.encoder.encode(packet, function(encodedPackets) {
+    if (rooms.length) {
+      for (var i = 0; i < rooms.length; i++) {
+        var room = self.rooms[rooms[i]];
+        if (!room) continue;
+        var sockets = room.sockets;
+        for (var id in sockets) {
+          if (sockets.hasOwnProperty(id)) {
+            if (ids[id] || ~except.indexOf(id)) continue;
+            socket = self.nsp.connected[id];
+            if (socket) {
+              socket.packet(encodedPackets, packetOpts);
+              ids[id] = true;
+            }
+          }
+        }
+      }
+    } else {
+      for (var id in self.sids) {
```

Comment:
__[jshint]__
* 'id' is already defined.

<!-- Comment by [SideCI](https://sideci.com) -->